projection.go

```diff
@@ -3,41 +3,72 @@ package dogma
 	Configure(c ProjectionConfigurer)
 
-	// HandleEvent handles a domain event message that has been routed to this
-	// handler.
+	// HandleEvent updates the projection to reflect the occurrence of event
+	// message m.
+	//
+	// If nil is returned, the projection has been updated successfully.
+	//
+	// If an error is returned, the projection SHOULD be left in the state it
+	// was before HandleEvent() was called.
+	//
+	// The engine SHOULD provide "at-least-once" delivery gaurantees to the
+	// handler. That is, the engine should call HandleEvent() with the same
+	// event message until it a nil error is returned.
+	//
+	// The engine MUST NOT call HandleEvent() with any message of a type that
+	// has not been routed to this handler by a prior call to Configure(). If
+	// any such message is passed, the implementation MUST panic with the
+	// UnexpectedMessage value.
 	//
-	// s provides access to the operations available within the scope of handling
-	// m.
+	// The engine MAY provide gaurantees about the order in which event messages
+	// will be passed to HandleEvent(), however in the interest of engine
+	// portability the implementation SHOULD NOT assume that HandleEvent() will
+	// be called with events in the same order that they were recorded.
 	//
-	// It panics with the UnexpectedMessage value if m is not one of the event
-	// types that is routed to this handler via Configure().
+	// The engine MAY call HandleEvent() from multiple goroutines concurrently.
 	HandleEvent(ctx context.Context, s ProjectionEventScope, m Message) error
 }
 
-// ProjectionConfigurer is an interface implemented by the engine and used
-// by the application to configure options related to a ProjectionMessageHandler.
+// ProjectionConfigurer is an interface implemented by the engine and used by
+// the application to configure options related to a ProjectionMessageHandler.
 //
 // It is passed to ProjectionMessageHandler.Configure(), typically upon
 // initialization of the engine.
 //
 // In the context of this interface, "the handler" refers to the handler on
 // which Configure() has been called.
 type ProjectionConfigurer interface {
-	// Name sets the name of the handler. Each handler within an application must
-	// have a unique name.
+	// Name sets the name of the handler.
+	//
+	// It MUST be called exactly once within a single call to Configure().
+	//
+	// Each handler within an application MUST have a unique, non-empty name.
 	Name(n string)
 
-	// RouteEventType configures the engine to route events of the same type as m
-	// to the handler.
+	// RouteEventType configures the engine to route event messages of the same
+	// type as m to the handler.
+	//
+	// It MUST be called at least once within a call to Configure(). It MUST NOT
+	// be called more than once with an event message of the same type.
+	//
+	// Multiple handlers within a single application MAY receive event messages
+	// of the same type.
```

Comment:
Does it mean to say multiple instances of ProjectionMessageHandler implementations MAY receive event of the same type?

Reply:
It might also be a process, not just another projection. I didn't feel it necessary to specify as these are the message handlers that can receive events at all.